### DISCUSSION TITLE
Ability to override Zend library

### DISCUSSION THREAD

**Discussion by henrykbrzoska** [OpenMage/magento-lts#3462] [Ideas]
### Description (*)
I would like to request an enhancement to be able override classes from `shardj/zf1-future` library.

### Expected behavior (*)
Not getting 
Fatal error: Cannot declare class Zend_XXX_YYY_ZZZ, because the name is already in use in /var/www/html/htdocs/vendor/shardj/zf1-future/library/Zend/XXX/YYY/ZZZ.php

### Benefits
Save the world.

### Additional information
as an example:
create app/code/local/Zend/Locale/Format.php
Comment out all require_once in file(because this classes are loaded). So was in lib/Zend before taking it from composer
change some item qty in cart
this will use Zend_Filter_LocalizedToNormalized
vendor/shardj/zf1-future/library/Zend/Filter/LocalizedToNormalized.php:30 throws
Fatal error: Cannot declare class Zend_Locale_Format, because the name is already in use in /var/www/html/htdocs/vendor/shardj/zf1-future/library/Zend/Locale/Format.php

**Comment by fballiano** [comment]
It would be nice, I don't know at the moment ho we could achieve it tho, any ideas?

**Reply by Sdfendor** [reply]
In our shop we override one smaller aspect of it via composer patch. But it's quite a bit of manual work and for larger rewrites it doesn't feels smooth.

**Reply by pquerner** [reply]
Can you share a sample composer patch, perhaps?
(No code changes, just change a word or something, just so that I see how exactly it works)

**Reply by Sdfendor** [reply]
While I could share my specific example, I rather first share the package I used to help me with creating this patch: https://github.com/symplify/vendor-patches. With the documentation given, I think the process is not complex. You only need the composer package for patch creation not to apply/run the patches. The created patches are applied via https://github.com/cweagans/composer-patches (which is also a dependency in the OpenMage composer.json and used to apply some specific changes to `shardj/zf1-future`).

**Comment by pquerner** [comment]
Like I proposed in Discord, maybe the [composer feature to autoload a file](https://getcomposer.org/doc/04-schema.md#files) can be used. Not sure if that invalidates PSR0 but I dont see how. You simply come first, and the autoloader finds your file first, before zf1future comes along.

```
{
    "name": "your/magento-base",
    "autoload": {
      "files": ["your/lib/Zend/Locale/Format.php"]
    }
}
```

where this package requires `openmage/magento-lts` (composer.json is unfinished obviously)

(How we would do that in non-composer setups however.. I dont know)

**Reply by fballiano** [reply]
well, I think this is the cleanest way and it can be used already today :-)

**Reply by henrykbrzoska** [reply]
but it is not working

**Reply by pquerner** [reply]
True, can confirm.

composer.json

```
{
    "name": "your/magento-base",
    "require": {
        "openmage/magento-lts": "^19.5.0",
        "aydin-hassan/magento-core-composer-installer": "*"
    },
    "config": {
        "allow-plugins": {
            "cweagans/composer-patches": true,
            "magento-hackathon/magento-composer-installer": true,
            "aydin-hassan/magento-core-composer-installer": true
        }
    },
    "php": "8.1",
    "autoload": {
      "files": ["./my_folder/Zend/Locale/Format.php"]
    },
    "extra": {
        "magento-core-package-type": "magento-source",
        "magento-root-dir": "htdocs"
    }
}

```

`Format.php`

```
<?php

class Zend_Locale_Format
{
    public function _construct() {
        die("hello world");
    }
}
```

Init with ddev, set php version to 8.1, load the page. Should crash immediately.

(did a `composer dump -o`)

composer autoload files show the file in the `_files.php` file and also in the `_static.php` file - therefore should be loaded correctly - but doesnt somehow.

Error with xdebug enabled:

```
<br class="Apple-interchange-newline">

( ! ) Fatal error: Cannot declare class Zend_Locale_Format, because the name is already in use in /var/www/html/vendor/shardj/zf1-future/library/Zend/Locale/Format.php on line 35
--


1 | 0.0089 | 385128 | {main}( ) | .../index.php:0
2 | 0.0148 | 462040 | Mage::run( $code = '', $type = 'store', $options = ??? ) | .../index.php:55
3 | 0.0185 | 600328 | Mage_Core_Model_App->run( $params = ['scope_code' => '', 'scope_type' => 'store', 'options' => []] ) | .../Mage.php:742
4 | 0.1403 | 1437504 | Mage_Core_Model_Resource_Setup::applyAllDataUpdates( ) | .../App.php:362
5 | 0.1405 | 1439264 | Mage_Core_Model_Resource_Setup->applyDataUpdates( ) | .../Setup.php:259
6 | 0.1405 | 1439488 | Mage_Core_Model_Resource_Setup->_installData( $newVersion = '1.6.0.0.19.1.6' ) | .../Setup.php:278
7 | 0.1405 | 1439488 | Mage_Core_Model_Resource_Setup->_modifyResourceDb( $actionType = 'data-install', $fromVersion = '', $toVersion = '1.6.0.0.19.1.6' ) | .../Setup.php:379
8 | 0.1409 | 1442352 | include( '/var/www/html/htdocs/app/code/core/Mage/Catalog/data/catalog_setup/data-install-1.6.0.0.php ) | .../Setup.php:610
9 | 0.1421 | 1516312 | Mage_Core_Model_Abstract->load( $id = 1, $field = ??? ) | .../data-install-1.6.0.0.php:21
10 | 0.1463 | 1589464 | Mage_Catalog_Model_Resource_Abstract->load( $object = class Mage_Catalog_Model_Category { protected $_data = []; protected $_hasDataChanges = FALSE; protected $_origData = []; protected $_idFieldName = 'entity_id'; protected $_isDeleted = FALSE; protected $_oldFieldsMap = []; protected $_syncFieldsMap = []; protected $_dirty = NULL; protected $_eventPrefix = 'catalog_category'; protected $_eventObject = 'category'; protected $_resourceName = 'catalog/category'; protected $_resource = NULL; protected $_resourceCollectionName = 'catalog/category_collection'; protected $_cacheTag = 'catalog_category'; protected $_dataSaveAllowed = TRUE; protected $_isObjectNew = TRUE; protected $_defaultValues = []; protected $_storeValuesFlags = []; protected $_lockedAttributes = []; protected $_isDeleteable = TRUE; protected $_isReadonly = FALSE; protected $_useFlatResource = FALSE; private $_designAttributes = [0 => 'custom_design', 1 => 'custom_design_from', 2 => 'custom_design_to', 3 => 'page_layout', 4 => 'custom_layout_update', 5 => 'custom_apply_to_products']; protected $_treeModel = NULL; protected $_urlModel = NULL }, $entityId = 1, $attributes = NULL ) | .../Abstract.php:234
11 | 0.1463 | 1589464 | Mage_Eav_Model_Entity_Abstract->load( $object = class Mage_Catalog_Model_Category { protected $_data = []; protected $_hasDataChanges = FALSE; protected $_origData = []; protected $_idFieldName = 'entity_id'; protected $_isDeleted = FALSE; protected $_oldFieldsMap = []; protected $_syncFieldsMap = []; protected $_dirty = NULL; protected $_eventPrefix = 'catalog_category'; protected $_eventObject = 'category'; protected $_resourceName = 'catalog/category'; protected $_resource = NULL; protected $_resourceCollectionName = 'catalog/category_collection'; protected $_cacheTag = 'catalog_category'; protected $_dataSaveAllowed = TRUE; protected $_isObjectNew = TRUE; protected $_defaultValues = []; protected $_storeValuesFlags = []; protected $_lockedAttributes = []; protected $_isDeleteable = TRUE; protected $_isReadonly = FALSE; protected $_useFlatResource = FALSE; private $_designAttributes = [0 => 'custom_design', 1 => 'custom_design_from', 2 => 'custom_design_to', 3 => 'page_layout', 4 => 'custom_layout_update', 5 => 'custom_apply_to_products']; protected $_treeModel = NULL; protected $_urlModel = NULL }, $entityId = 1, $attributes = NULL ) | .../Abstract.php:737
12 | 0.1497 | 1860320 | Mage_Eav_Model_Entity_Abstract->_afterLoad( $object = class Mage_Catalog_Model_Category { protected $_data = []; protected $_hasDataChanges = FALSE; protected $_origData = []; protected $_idFieldName = 'entity_id'; protected $_isDeleted = FALSE; protected $_oldFieldsMap = []; protected $_syncFieldsMap = []; protected $_dirty = NULL; protected $_eventPrefix = 'catalog_category'; protected $_eventObject = 'category'; protected $_resourceName = 'catalog/category'; protected $_resource = NULL; protected $_resourceCollectionName = 'catalog/category_collection'; protected $_cacheTag = 'catalog_category'; protected $_dataSaveAllowed = TRUE; protected $_isObjectNew = TRUE; protected $_defaultValues = []; protected $_storeValuesFlags = []; protected $_lockedAttributes = []; protected $_isDeleteable = TRUE; protected $_isReadonly = FALSE; protected $_useFlatResource = FALSE; private $_designAttributes = [0 => 'custom_design', 1 => 'custom_design_from', 2 => 'custom_design_to', 3 => 'page_layout', 4 => 'custom_layout_update', 5 => 'custom_apply_to_products']; protected $_treeModel = NULL; protected $_urlModel = NULL } ) | .../Abstract.php:944
13 | 0.1497 | 1860696 | Mage_Eav_Model_Entity_Abstract->walkAttributes( $partMethod = 'backend/afterLoad', $args = [0 => class Mage_Catalog_Model_Category { protected $_data = [...]; protected $_hasDataChanges = FALSE; protected $_origData = [...]; protected $_idFieldName = 'entity_id'; protected $_isDeleted = FALSE; protected $_oldFieldsMap = [...]; protected $_syncFieldsMap = [...]; protected $_dirty = NULL; protected $_eventPrefix = 'catalog_category'; protected $_eventObject = 'category'; protected $_resourceName = 'catalog/category'; protected $_resource = NULL; protected $_resourceCollectionName = 'catalog/category_collection'; protected $_cacheTag = 'catalog_category'; protected $_dataSaveAllowed = TRUE; protected $_isObjectNew = TRUE; protected $_defaultValues = [...]; protected $_storeValuesFlags = [...]; protected $_lockedAttributes = [...]; protected $_isDeleteable = TRUE; protected $_isReadonly = FALSE; protected $_useFlatResource = FALSE; private $_designAttributes = [...]; protected $_treeModel = NULL; protected $_urlModel = NULL }] ) | .../Abstract.php:1617
14 | 0.1502 | 1862768 | Mage_Eav_Model_Entity_Attribute_Backend_Time_Created->afterLoad( $object = class Mage_Catalog_Model_Category { protected $_data = []; protected $_hasDataChanges = FALSE; protected $_origData = []; protected $_idFieldName = 'entity_id'; protected $_isDeleted = FALSE; protected $_oldFieldsMap = []; protected $_syncFieldsMap = []; protected $_dirty = NULL; protected $_eventPrefix = 'catalog_category'; protected $_eventObject = 'category'; protected $_resourceName = 'catalog/category'; protected $_resource = NULL; protected $_resourceCollectionName = 'catalog/category_collection'; protected $_cacheTag = 'catalog_category'; protected $_dataSaveAllowed = TRUE; protected $_isObjectNew = TRUE; protected $_defaultValues = []; protected $_storeValuesFlags = []; protected $_lockedAttributes = []; protected $_isDeleteable = TRUE; protected $_isReadonly = FALSE; protected $_useFlatResource = FALSE; private $_designAttributes = [0 => 'custom_design', 1 => 'custom_design_from', 2 => 'custom_design_to', 3 => 'page_layout', 4 => 'custom_layout_update', 5 => 'custom_apply_to_products']; protected $_treeModel = NULL; protected $_urlModel = NULL } ) | .../Abstract.php:641
15 | 0.1502 | 1863040 | Mage_Core_Model_Locale->storeDate( $store = NULL, $date = NULL, $includeTime = TRUE, $format = NULL ) | .../Created.php:74
16 | 0.1502 | 1863088 | Composer\Autoload\ClassLoader->loadClass( $class = 'Zend_Date' ) | .../Locale.php:548
17 | 0.1502 | 1863200 | Composer\Autoload\{closure:/var/www/html/vendor/composer/ClassLoader.php:575-577}( $file = '/var/www/html/vendor/composer/../shardj/zf1-future/library/Zend/Date.php' ) | .../ClassLoader.php:427
18 | 0.1535 | 1929096 | include( '/var/www/html/vendor/shardj/zf1-future/library/Zend/Date.php ) | .../ClassLoader.php:576
19 | 0.1565 | 1929648 | require_once( '/var/www/html/vendor/shardj/zf1-future/library/Zend/Locale/Format.php ) | .../Date.php:27


```

**Comment by pquerner** [comment]
Otherwise we put this in the Mage.php to allow special overwrites for Zend, which can be extended in the future?
(Or better yet, since Mage is final, somewhere else which is better to oberwrite - or outsource to XML Struct)

```
spl_autoload_register(function ($class) {
    // if Zend and exists in app/local, then include from there, else ask the composer autoloader (exit here)
}, true, true); 
```

**Reply by pquerner** [reply]
https://github.com/OpenMage/magento-lts/blob/b1a879c75e07d88a92dbd8345786ee080923d52c/app/Mage.php#L57-L59

Can we load our overwritten Zend classes here? And when the original `require_once` comes along our class is simply already loaded?

(If its an easy Zend class, with not too many `require_once` itself)

**Comment by henrykbrzoska** [comment]
> Otherwise we put this in the Mage.php to allow special overwrites for Zend, which can be extended in the future? (Or better yet, since Mage is final, somewhere else which is better to oberwrite - or outsource to XML Struct)
> 
> ```
> spl_autoload_register(function ($class) {
>     // if Zend and exists in app/local, then include from there, else ask the composer autoloader (exit here)
> }, true, true); 
> ```
it will not work because other parts of zend loads overriden class second time
![image](https://github.com/OpenMage/magento-lts/assets/4395216/d70d7f9c-9d19-435f-b41b-e3ce0ec51129)

this shardj/zf1-future is not compatible with old lib/Zend because library was loaded with require_once commented out
old one:
![image](https://github.com/OpenMage/magento-lts/assets/4395216/7ebc9bd5-9cbc-4e2a-b9e8-efdae2b481c8)

the only one way to change something in Zend/Locale/Format.php is to override every class which have require_once 'Zend/Locale/Format.php';
than override every class which have class which require Zend/Locale/Format.php
repeat with every class which have class which have class which require Zend/Locale/Format.php

**Reply by pquerner** [reply]
So you are saying you tell the autoloader that it can find the class `Zend_Locale_Format` at `/some/where/Zend/Locale/Format.php` and then when the class `Zend_Filter_LocalizedToNormalized` comes along, it will retrigger the autoloader and ignores your `/some/where/Zend/Locale/Format.php` but instead goes into `vendor/zf1future/.../Format.php` ?
Why?

**Reply by henrykbrzoska** [reply]
but vendor/zf1future/.../Format.php is not loaded by autoloader. It is loaded by vendor/shardj/zf1-future/library/Zend/Filter/LocalizedToNormalized.php line 30
there we have 
`require_once 'Zend/Locale/Format.php'; 
`

The main issue is that previously(in lib/Zend) this line was commented out(see second screenshot), and here(zf1future) is not commented out.

**Reply by henrykbrzoska** [reply]
maby we need a fork of zf1future where we make something like this
`if(!class_exists(Zend_Locale_Format)){
    require_once 'Zend/Locale/Format.php';
}`
with every require_once

**Reply by pquerner** [reply]
We should ask the devs of zf1future if its really needed to have these `require_once` statements in there in the age of composer ...
Probably to not poke the bear and leave it as is.

**Reply by henrykbrzoska** [reply]
from what i remember that was the one of main reasons that pushed us/them/IBM/'php world' to use psr-4 and namespaces

**Comment by pquerner** [comment]
@henrykbrzoska Did you continue to work on a solution for this?
Best last guess would be to use a fork of zf1future and make your changes there. Meaning you have to keep up with the changes yourself, instead of OM releases.

**Comment by elidrissidev** [comment]
The best thing would be for zf1-future to make use of the composer autoloader and remove all the `require_once` statements from the code, but alas shardj/zf1-future#348...

**Reply by pquerner** [reply]
https://github.com/zf1s/zf1 was given as different approach. Maybe we can trick composer.json to use https://github.com/zf1s/zf1 for zf1future instead - and then we can use composer autoloader to load our own files instead?

**Reply by elidrissidev** [reply]
`zf1s` doesn't seem to have much activity as `zf1-future` so idk about that. Last solution would be to create a patch to comment out `require_once` calls in vendor and call it a day.

**Reply by sreichel** [reply]
+1 for `zf1s` .... already mentioned it here #3199, b/c we dont need all from Zend framework. We could fork only the parts, that are required. 

(maybe think about to modulize OM ... like Mage_Poll, Mage_Backup. Already started it, but then Mage_Adminhtml classes have to be forwarded to new classes inside the module.)

**Reply by pquerner** [reply]
@fballiano Can we switch the zf1 repo? They gave a hard "no" to change the require_once statements and OpenMage moves into a more up-to-date light anyways, so why keep the old habits in zf1 while being "all new" at the front?

This comes to mind for this, lol:

![image](https://i.kym-cdn.com/photos/images/newsfeed/001/399/018/31f.gif)

**Reply by Flyingmana** [reply]
how does the alternatives compare in regards of number of maintainers, consumers, and number of fixes and patches?(and also regarding security fixes)

And, what would be the real benefits for us?

Users of openmage can easily use composer to configure an alternative already, so the change on our side is only a minimal benefit in this regard.
What other benefits are there for us as maintainers and for us as project?

**Reply by pquerner** [reply]
Well.. OM is moving to a PHP 8.x compat, while zf1future only promises 8.1 compat.
Benefit is using composer autoloader, which enables to mitigate this very issue thats being discussed here.
(the other linked zf1 repo brings up to 8.3 compat)

In terms of bugs / performance things how these two differs, I dont know.

> Users of openmage can easily use composer to configure an alternative already

How? Do you have an example composer.json file at hand?

> What other benefits are there for us as maintainers and for us as project?

That is difficult to answer, because I dont know how these 2 repos differ. Just that the other one does not rely on require_onces in the codebase and the current in use is strict about "this is for legacy projects, where require_once had to be used for the lack of a proper autoloader".

**Reply by fballiano** [reply]
with all the patches we provided to ZF1F it would be impossible to retest everything

**Reply by pquerner** [reply]
Ok, so how is this fixable now - and also in terms of PHP compat version?

**Reply by fballiano** [reply]
I don't know if that's fixable at all, to fork zf1f it would be a nightmare, who would maintain that?

**Reply by pquerner** [reply]
But it was a feature to copy a Zend file to codepool "local" and it would work. Well.. who did that anyways.. ;)

> I don't know if that's fixable at all, to fork zf1f it would be a nightmare, who would maintain that?

Looking at the releases they are frequently in recent times, but not overwhelming. And to get rid of all `require_once` calls shouldnt be a nightmare too. Thats a simple `sed` script at most.

Then running `composer dump -o` - or rather telling the OpenMage composer.json's autoloader config where to find scripts (too) and it would be done?

**Reply by fballiano** [reply]
could you create a patch to zf1f and then apply it with composer patch?

**Reply by pquerner** [reply]
I'll look into it.

**Reply by Flyingmana** [reply]
>  while zf1future only promises 8.1 compat

and they declined to provide higher compat in the future?

> > Users of openmage can easily use composer to configure an alternative already
> 
> How? Do you have an example composer.json file at hand?

not sure, its a few years since I last did something in this direction.
maybe its https://getcomposer.org/doc/articles/aliases.md#require-inline-alias


> > What other benefits are there for us as maintainers and for us as project?
> 
> That is difficult to answer, because I dont know how these 2 repos differ. Just that the other one does not rely on require_onces in the codebase and the current in use is strict about "this is for legacy projects, where require_once had to be used for the lack of a proper autoloader".

For us this is important, because we could easily switch, but then someone else comes with another reason why zf1f has something they think is more important, so we will switch back again.
Having the full comparison enables us to do a longterm decision which protects us from a situation where we need to change course only a year later.

**Reply by pquerner** [reply]
> and they declined to provide higher compat in the future?

its how it is as of today. i've not read their issues or pr chats to find an answer to that question.

**Reply by Flyingmana** [reply]
they seem to accept patches for it
https://github.com/search?q=repo%3AShardj%2Fzf1-future+php8.2&type=pullrequests
but there is not much activity.

**Reply by Flyingmana** [reply]
had a small look, seems to be good enough of an alternative, and having more activity for new php versions is a good reason.
Although, the php8.3 changes seem rather small https://github.com/zf1s/zf1/pull/182/files

I also make a note ( https://github.com/OpenMage/organizational/issues/14 ) to document the decision with its most relevant points when it moves close to be finished (in case we really proceed with this)

**Reply by fballiano** [reply]
we should make sure that zf1s has this patches that we provided to zf1f:
https://github.com/shardj/zf1-future/pulls?q=is%3Apr+author%3Afballiano+is%3Aclosed
https://github.com/shardj/zf1-future/pulls?q=is%3Apr+author%3Asreichel+is%3Aclosed

I don't see the switch as a benefit in this moment, I don't have any problem with zf1f on php8.3 and we have many PRs that they merged so they've not been closed to implementing changes

**Comment by elidrissidev** [comment]
I think I found a working solution, please check https://github.com/OpenMage/magento-lts/pull/3989.

**Reply by sreichel** [reply]
It works for zf1s, but not for zf1-future ... :(

**Comment by luigifab** [comment]
Note: before zf1-future, with the included zend files, there were a script to comment all `require_once`.

**Reply by pquerner** [reply]
Cant find it. Where?

**Reply by luigifab** [reply]
[Here](https://github.com/OpenMage/magento-lts/blob/v19.4.0/lib/Zend/replace_recursive.php)

**Reply by luigifab** [reply]
Is zf1-future installed via composer still works after running the script?

**Reply by sreichel** [reply]
@luigifab it does, but comment `require_once` does not solve the problem :(
